rabbitmq/client.go

```diff
@@ -0,0 +1,109 @@
+package rabbitmq
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/palantir/stacktrace"
+
+	"github.com/streadway/amqp"
+)
+
+// ReconnectDelay ...
+const ReconnectDelay = 2
+const RetryAttempts = 5
+
+type RabbitMqClientConfig struct {
+	Host                  string
+	Port                  int
+	Username              string
+	Password              string
+	Metrics               Metric
+	ConnectRetryAttempts  int
+	InitialReconnectDelay time.Duration
+}
+
+type RabbitMQClient struct {
+	amqpURI string
+	conn    *amqp.Connection
+	channel *amqp.Channel
+	metric  Metric
+}
+
+func NewRabbitMQClient(config RabbitMqClientConfig) (*RabbitMQClient, error) {
+	amqpURI := fmt.Sprintf(
+		"amqp://%s:%s@%s:%d/",
+		config.Username,
+		config.Password,
+		config.Host,
+		config.Port,
+	)
+
+	client := &RabbitMQClient{
+		amqpURI: amqpURI,
+		metric:  config.Metrics,
+	}
+
+	err := client.establishConnection()
+	if err != nil {
+		return nil, err
+	}
+
+	err = client.establishChannel()
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+func (c *RabbitMQClient) establishConnection() error {
```

Comment:
Merge this establish connection logic as part of the constructor, unless we need a case where `establishConnection` may be called multiple times.

I think there's no such case.